controller/server/main.go

```diff
@@ -14,25 +14,129 @@
+	d := &deploy.Deployment{}
+	switch kind := req.ClusterSpec.(type) {
+	case *cpb.CreateClusterRequest_Kind:
+		d.Cluster = &deploy.KindSpec{
+			Name:                     req.GetKind().Name,
+			Recycle:                  req.GetKind().Recycle,
+			Version:                  req.GetKind().Version,
+			Image:                    req.GetKind().Image,
+			Retain:                   req.GetKind().Retain,
+			GoogleArtifactRegistries: req.GetKind().GoogleArtifactRegistries,
+		}
+	default:
+		return nil, fmt.Errorf("cluster type not supported: %T", kind)
+	}
+	switch metallb := req.IngressSpec.(type) {
+	case *cpb.CreateClusterRequest_Metallb:
+		l := &deploy.MetalLBSpec{}
+		if req.GetMetallb().ManifestDir != "" {
+			defaultMetallbManifestDir = req.GetMetallb().ManifestDir
+
+		}
+		p, err := validatePath(defaultMetallbManifestDir)
+		if err != nil {
+			return nil, fmt.Errorf("failed to validate path %q", defaultMetallbManifestDir)
+		}
+		l.ManifestDir = p
+		l.IPCount = int(req.GetMetallb().IpCount)
+		l.Version = req.GetMetallb().Version
+		d.Ingress = l
+	default:
+		return nil, fmt.Errorf("ingress spec not supported: %T", metallb)
+	}
+	switch meshnet := req.CniSpec.(type) {
+	case *cpb.CreateClusterRequest_Meshnet:
+		m := &deploy.MeshnetSpec{}
+		if req.GetMeshnet().ManifestDir != "" {
+			defaultMeshnetManifestDir = req.GetMeshnet().ManifestDir
+		}
+		p, err := validatePath(defaultMeshnetManifestDir)
+		if err != nil {
+			return nil, fmt.Errorf("failed to validate path %q", defaultMeshnetManifestDir)
+		}
+		m.Image = req.GetMeshnet().Image
+		m.ManifestDir = p
+		d.CNI = m
+	default:
+		return nil, fmt.Errorf("cni type not supported: %T", meshnet)
+	}
+	return d, nil
+
+}
+
+func (s *server) CreateCluster(ctx context.Context, req *cpb.CreateClusterRequest) (*cpb.CreateClusterResponse, error) {
+	log.Infof("Creating new cluster")
+	d, err := newDeployment(req)
+	if err != nil {
+		return nil, err
+	}
+	if err := d.Deploy(ctx, defaultKubeCfg); err != nil {
+		resp := &cpb.CreateClusterResponse{
+			Name:  req.GetKind().Name,
+			State: cpb.ClusterState_CLUSTER_STATE_ERROR,
+		}
+		return resp, fmt.Errorf("failed to deploy cluster: %s", err)
+	}
+	log.Infof("Cluster: %q deployed and ready for topology.", req.GetKind().Name)
+	resp := &cpb.CreateClusterResponse{
+		Name:  req.GetKind().Name,
+		State: cpb.ClusterState_CLUSTER_STATE_RUNNING,
+	}
+	return resp, nil
+}
+
+func validatePath(path string) (string, error) {
+	path, err := filepath.Abs(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to evaluate absolute path: %q", path)
+	}
+	if _, err := os.Stat(path); err != nil {
+		return "", fmt.Errorf("path %q does not exist", path)
+	}
+	return path, nil
+}
+
+func Init() {
+	if home := homedir.HomeDir(); home != "" {
+		defaultKubeCfg = filepath.Join(home, ".kube", "config")
+		defaultMeshnetManifestDir = filepath.Join(home, "kne/manifests/meshnet/base")
+		defaultMetallbManifestDir = filepath.Join(home, "kne/manifests/metallb")
+	}
+
```

Comment:
There are extra newlines here and elsewhere that are uneccesaary

Reply:
done